Source/Extensions/CAShapeLayerExtension.swift

```diff
@@ -0,0 +1,151 @@
+//
+//  CAShapeLayerExtension.swift
+//  
+//
+//  Created by Ehab Alsharkawy on 14/11/2022.
+//
+
+import UIKit
+
+extension CAShapeLayer {
+
+    /// Add Edges and corners with style
+    /// - Parameter style: Protocol that provide which edge or corner will be added with style
+    func updateEdgesAndCorners(with style: ElementBorderStyle) {
+
+        // if path not nil then it's already added so no need to draw lines or curves again
+        guard self.path == nil else { return }
+
+        // always fill is clear color
+        fillColor = UIColor.clear.cgColor
+
+        lineWidth = style.borderWidth
+        let rect = bounds.insetBy(dx: style.cornerRadius / 2,
+                                  dy: style.cornerRadius / 2)
+        let path = UIBezierPath()
+        updateBorders(with: path, rect: rect, style: style)
+        updateCorners(with: path, rect: rect, style: style)
+        self.path = path.cgPath
+    }
+
+    /*
+     (0,0)  c-----------c  (100,0)
+            |           |
+            |           |
+            |           |
+     (0,30) c-----------c  (100,30)
+
+     origin = (0,0)
+     minx = 0
+     minY = 0
+     maxX = 100
+     maxY = 30
+     */
+
+    /// Add corners with style
+    /// - Parameters:
+    ///   - path: UIBezierPath that will be added
+    ///   - rect: CGRect of current view
+    ///   - style: Protocol that provide which edge will be added
+    private func updateBorders(with path: UIBezierPath, rect: CGRect, style: ElementBorderStyle) {
+        guard let edges = style.edges else { return }
+
+        //  Bottom Left corner Radius
+        let isBottomLeftCornerRequired = style.corners?.contains(.bottomLeft) ?? false
+        let bottomLeftCornerRadius: CGFloat = isBottomLeftCornerRequired ? style.cornerRadius : 0
+
+        // Top Left corner Radius
+        let isTopLeftCornerRequired = style.corners?.contains(.topLeft) ?? false
+        let topLeftCornerRadius: CGFloat = isTopLeftCornerRequired ? style.cornerRadius : 0
+        print(topLeftCornerRadius)
+
+        //  Top Right corner Radius
+        let isTopRightCornerRequired = style.corners?.contains(.topRight) ?? false
+        let topRightCornerRadius = isTopRightCornerRequired ? style.cornerRadius : 0
+
+        // Bottom Right corner Radius
+        let isBottomRightCornerRequired = style.corners?.contains(.bottomRight) ?? false
+        let bottomRightCornerRadius = isBottomRightCornerRequired ? style.cornerRadius : 0
+
+        // left line
+        if edges.contains(.left) {
+            path.drawLine(startX: rect.minX,
+                          startY: rect.minY + topLeftCornerRadius,
```

Comment:
For these corner radius offsets, I really wonder whether they are full offset on both lines, or half of offset on both sides.

If you have a setup to easily check, could you try a standard border with a corner radius of something big, like 50, then try the same with this approach and check whether they look the same?